src/celeritas/geo/GeoMaterialParams.cc

```diff
@@ -28,6 +28,211 @@
 
 namespace celeritas
 {
+namespace
+{
+//---------------------------------------------------------------------------//
+using MapLabelMatId = std::unordered_map<Label, MaterialId>;
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct a label -> material map from the input.
+ *
+ * The input is effectively an "unzipped" unordered list of (volume label,
+ * material pairs).
+ */
+MapLabelMatId build_label_map(MaterialParams const& mat_params,
+                              std::vector<Label>&& labels,
+                              std::vector<MaterialId> const& materials)
+{
+    CELER_EXPECT(materials.size() == labels.size());
+    CELER_EXPECT(std::all_of(
+        materials.begin(), materials.end(), [&mat_params](MaterialId m) {
+            return !m || m < mat_params.num_materials();
+        }));
+
+    std::unordered_map<Label, MaterialId> lab_to_id;
+    std::set<Label> duplicates;
+
+    // Remap materials to volume IDs using given volume names:
+    // build a map of volume name -> matid
+    for (auto idx : range(materials.size()))
+    {
+        if (!materials[idx])
+        {
+            // Skip volumes without matids
+            continue;
+        }
+
+        auto [prev, inserted]
+            = lab_to_id.insert({std::move(labels[idx]), materials[idx]});
+        if (!inserted)
+        {
+            duplicates.insert(prev->first);
+        }
+    }
+    CELER_VALIDATE(duplicates.empty(),
+                   << "geo/material coupling specified duplicate "
+                      "volume names: \""
+                   << join(duplicates.begin(), duplicates.end(), "\", \"")
+                   << '"');
+    return lab_to_id;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Find a material ID from a volume ID.
+ */
+class MaterialFinder
+{
+  public:
+    MaterialFinder(GeoParams const& geo, MapLabelMatId const& materials)
+        : geo_{geo}, materials_{materials}
+    {
+    }
+
+    MaterialId operator()(VolumeId const& volume_id)
+    {
+        Label const& vol_label = geo_.id_to_label(volume_id);
+
+        // Hopefully user-provided and geo-provided volume labels match exactly
+        if (auto iter = materials_.find(vol_label); iter != materials_.end())
+        {
+            return iter->second;
+        }
+
+        if (mat_labels_.empty())
+        {
+            // Build user volume -> material mapping
+            this->build_mat_labels();
```

Comment:
This seems a somewhat expensive operation.  Is it done only once per process or more often?

Reply:
Once per process. It's cached here.

Reply:
Ok.  So per convention this is okay but this code is not thread safe (and it does not claim to be since this is not a const function (but semantically it should be, shouldn't it).

Do we have a handle on the list of class/object that are shared amongst threads but are not thread safe? 

Actually the call to `build_mat_labels` is delayed until the first call to `operator ()`, do we really have a guarantee (in all problem setups) that there will be such a call before this is used in multi-thread mode?

Reply:
This is literally an implementation detail called locally by a constructor. It doesn't exceed the lifetime of a local function. It doesn't access any shared resources. It is therefore thread safe.

Reply:
Fair enough.  It may or may not be useful to mark a little more clearly (yes, I have finally noticed it is declared inside the source file) that this is an internal class (i.e "literally an implementation detail"), for example by putting into a 'detail' or 'unnamed' namespace.

Reply:
@pcanal It *is* in an anonymous namespace.

Reply:
Perfect then (and sorry; it is not obvious from the github output :) ).